File: Agora-Rtm-Tutorial-macOS-Objective-C/Agora-Rtm-Tutorial/PeerChannelViewController.m
//
//  PeerChannelViewController.m
//  Agora-Rtm-Tutorial-Mac
//
//  Created by CavanSu on 2019/2/22.
//  Copyright © 2019 Agora. All rights reserved.
//

#import "PeerChannelViewController.h"
#import "ChatViewController.h"

@interface PeerChannelViewController () <ChatVCDelegate>
@property (weak, nonatomic) IBOutlet NSTextField *peerTextField;
@property (weak, nonatomic) IBOutlet NSTextField *channelTextField;

@property (nonatomic, assign) ChatMode selectedMode;
@end

@implementation PeerChannelViewController

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (IBAction)doChatPressed:(NSButton *)sender {
    NSString *peer = self.peerTextField.stringValue;
    
    if (!peer.length) {
        return;
    }
    
    ChatMode mode;
    mode.type = ChatTypePeer;
    mode.name = peer;
    
    self.selectedMode = mode;
    
    [self performSegueWithIdentifier:@"peerChannelToChat" sender:nil];
}

- (IBAction)doJoinPressed:(NSButton *)sender {
    NSString *channel = self.channelTextField.stringValue;
    
    if (!channel.length) {
        return;
    }
    
    ChatMode mode;
    mode.type = ChatTypeGroup;
    mode.name = channel;
    
    self.selectedMode = mode;
    
    [self performSegueWithIdentifier:@"peerChannelToChat" sender:nil];
}

- (IBAction)doBackPressed:(NSButton *)sender {
    if (self.delegate && [self.delegate respondsToSelector:@selector(peerChannelVCWillClose:)]) {
        [self.delegate peerChannelVCWillClose:self];
    }
}

- (void)prepareForSegue:(NSStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"peerChannelToChat"]) {
        ChatViewController *chatVC = (ChatViewController *)segue.destinationController;
        chatVC.mode = self.selectedMode;
        chatVC.delegate = self;
    }
}

- (void)chatVCWillClose:(ChatViewController *)vc {
    vc.view.window.contentViewController = self;
}

@end

File: Agora-RTM-Tutorial-macOS/Agora-RTM-Tutorial/PeerChannelViewController.swift
//
//  PeerChannelViewController.swift
//  Agora-Rtm-Tutorial-Mac
//
//  Created by CavanSu on 2019/1/31.
//  Copyright © 2019 Agora. All rights reserved.
//

import Cocoa
import AgoraRtmKit

protocol PeerChannelVCDelegate: NSObjectProtocol {
    func peerChannelVCWillClose(_ vc: PeerChannelViewController)
}

class PeerChannelViewController: NSViewController, ShowAlertProtocol {
    @IBOutlet weak var peerTextField: NSTextField!
    @IBOutlet weak var channelTextField: NSTextField!
    
    var delegate: PeerChannelVCDelegate?
    
    override func viewWillAppear() {
        AgoraRtm.updateKit(delegate: self)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func prepare(for segue: NSStoryboardSegue, sender: Any?) {
        guard let _ = segue.identifier, let type = sender as? ChatType else {
            return
        }
        let chatVC = segue.destinationController as! ChatViewController
        chatVC.type = type
        chatVC.delegate = self
    }
    
    @IBAction func doChatPressed(_ sender: NSButton) {
        let peer = peerTextField.stringValue
        guard peer.count > 0 else {
            return
        }
        let type = ChatType.peer(peer)
        performSegue(withIdentifier: "peerChannelToChat", sender: type)
    }
    
    @IBAction func doJoinPressed(_ sender: NSButton) {
        let channel = channelTextField.stringValue
        guard  channel.count > 0 else {
            return
        }
        let type = ChatType.group(channel)
        performSegue(withIdentifier: "peerChannelToChat", sender: type)
    }
    
    @IBAction func doBackPressed(_ sender: NSButton) {
        self.delegate?.peerChannelVCWillClose(self)
    }
}

extension PeerChannelViewController: AgoraRtmDelegate {
    func rtmKit(_ kit: AgoraRtmKit, connectionStateChanged state: AgoraRtmConnectionState, reason: AgoraRtmConnectionChangeReason) {
        showAlert("connection state changed: \(state.rawValue)") { [weak self] (_) in
            if reason == .remoteLogin, let strongSelf = self {
                strongSelf.delegate?.peerChannelVCWillClose(strongSelf)
            }
        }
    }
    
    // Receive one to one offline messages
    func rtmKit(_ kit: AgoraRtmKit, messageReceived message: AgoraRtmMessage, fromPeer peerId: String) {
        AgoraRtm.add(offlineMessage: message, from: peerId)
    }
    
    func rtmKit(_ kit: AgoraRtmKit, imageMessageReceived message: AgoraRtmImageMessage, fromPeer peerId: String) {
        AgoraRtm.add(offlineMessage: message, from: peerId)
    }
}

extension PeerChannelViewController: ChatVCDelegate {
    func chatVCWillClose(_ vc: ChatViewController) {
        vc.view.window?.contentViewController = self
    }
    
    func chatVCBackToRootVC(_ vc: ChatViewController) {
        self.delegate?.peerChannelVCWillClose(self)
    }
}
